### PR TITLE
Add support for Mumble positional audio

### LINKED ISSUE #1753: Add mumble positional Audio
http://wiki.mumble.info/wiki/Link

### PR DESCRIPTION
**Pull request by ShadowNinja**
Fixes #1753.
The overlay works, but I haven't actually tested the audio communication yet.

### CODE REVIEW

File: src/client/mumble.cpp [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+		obj = NULL;
+		return false;
+	}
+#else
+	std::string name = "/MumbleLink." + to_string(getuid());
+
+	fd = shm_open(name.c_str(), O_RDWR, S_IRUSR | S_IWUSR);
+
+	if (fd < 0) {
+		std::cout << "Failed to create SHM for mumble!" << std::endl;
```
Comment (est31) [inline src/client/mumble.cpp]: can this go over our logging system instead?

Comment (est31) [inline src/client/mumble.cpp]: especially important for windows users

File: src/client/mumble.cpp [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+
+void MumbleManager::step(const v3f &player_pos, const v3f &cam_pos, v3f cam_dir)
+{
+	if (lm == NULL)
+		return;
+
+	cam_dir.normalize();
+
+	// Todo: calculate from yaw and pitch
```
Comment (est31) [inline src/client/mumble.cpp]: I think this is already done, so perhaps remove this comment?

File: src/client/mumble.cpp [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+	}
+#endif
+
+	// Context is equal for players on the same server
+	// TODO: allow mods to append something to allow for teams
+	size_t len = MYMIN(sizeof(lm->context), context.size());
+	memcpy(lm->context, context.c_str(), len);
+	lm->context_len = len;
+
+	std::wstring wpn = narrow_to_wide(player_name);
```
Comment (est31) [inline src/client/mumble.cpp]: can you use `utf8_to_wide` instead?

Comment (ShadowNinja) [inline src/client/mumble.cpp]: I'm not sure how the player name is stored.  Is it stored as UTF-8 or the native encoding?

Comment (est31) [inline src/client/mumble.cpp]: @ShadowNinja its stored as wstring when entered into the formspec, then its converted UTF-8 in guiformspec.cpp, and throughout the engine UTF-8 is used. Only place its converted to widestring is for chat I think because there we send wstrings over the line atm. So most probably it will be in utf8. Not that it would matter, but i'd like to remove the narrow_to_wide function once we use utf-8 based chat.

Comment (est31) [inline src/client/mumble.cpp]: (both utf-8 and narrow would "work" here, as the player name is already restricted to a subset of ASCII), but see above, I'd like to remove narrow_to_wide.

**Comment by ShadowNinja** [comment]
This should be ready to use now.  I'm not very confident that the vector rotation lines are mathematically correct, but they seem to work...

File: src/filesys.cpp [hunk 4]
```diff
@@ -488,7 +489,7 @@ bool CopyFileContents(const std::string &source, const std::string &target)
 		}
 		if(ferror(targetfile)){
 			errorstream<<target<<": IO error: "
-					<<strerror(errno)<<std::endl;
+					<<porting::strerrno(errno)<<std::endl;
```
Comment (est31) [inline src/filesys.cpp]: Can you do this in a separate commit? If you do already, then don't squash before merging :)

File: src/porting.h [hunk 5]
```diff
@@ -60,7 +60,16 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 	#include <unistd.h>
 	#include <stdint.h> //for uintptr_t
 
-	#if (defined(linux) || defined(__linux) || defined(__GNU__)) && !defined(_GNU_SOURCE)
+	// Define some platform flags that require a lot of checks
```
Comment (est31) [inline src/porting.h]: here too. I agree that its useful, just its unrelated to the change, so would be best to keep it separate.

File: src/filesys.cpp (hunk 4 above)
Comment (ShadowNinja) [inline src/filesys.cpp]: Yes, it's seperate.

File: src/porting.h (hunk 5 above)
Comment (ShadowNinja) [inline src/porting.h]: They're seperate, see the "Commits" tab on the PR.

**Comment by ENDev15** [comment]
So how is this feature coming along? Mumble is great and it would be epic to have its full functionality available in Minetest.

**Comment by ShadowNinja** [comment]
@ENDev15: This needs to be tested and confirmed working.  And I'd like it if someone could confirm that the math for the top vectors is right.

**Comment by Elinvention** [comment]
I tried to compile Minetest with this PR on Linux, but I get compilation errors:

```
minetest/src/client/mumble.cpp:39:1: error: ‘MumbleManager’ does not name a type
 MumbleManager::~MumbleManager()
 ^
minetest/src/client/mumble.cpp:53:6: error: ‘MumbleManager’ has not been declared
 bool MumbleManager::start(const std::string &context,
      ^
minetest/src/client/mumble.cpp: In function ‘bool start(const string&, const string&)’:
minetest/src/client/mumble.cpp:77:2: error: ‘fd’ was not declared in this scope
  fd = shm_open(name.c_str(), O_RDWR, S_IRUSR | S_IWUSR);
  ^
minetest/src/client/mumble.cpp:85:2: error: ‘lm’ was not declared in this scope
  lm = static_cast<MumbleLinkedMemory *>(mmap(NULL, sizeof(lm),
  ^
minetest/src/client/mumble.cpp:85:19: error: ‘MumbleLinkedMemory’ does not name a type
  lm = static_cast<MumbleLinkedMemory *>(mmap(NULL, sizeof(lm),
                   ^
minetest/src/client/mumble.cpp:85:38: error: expected ‘>’ before ‘*’ token
  lm = static_cast<MumbleLinkedMemory *>(mmap(NULL, sizeof(lm),
                                      ^
minetest/src/client/mumble.cpp:85:38: error: expected ‘(’ before ‘*’ token
minetest/src/client/mumble.cpp:85:39: error: expected primary-expression before ‘>’ token
  lm = static_cast<MumbleLinkedMemory *>(mmap(NULL, sizeof(lm),
                                       ^
minetest/src/client/mumble.cpp:86:47: error: expected ‘)’ before ‘;’ token
    PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0));
                                               ^
minetest/src/client/mumble.cpp: At global scope:
minetest/src/client/mumble.cpp:116:6: error: ‘MumbleManager’ has not been declared
 void MumbleManager::step(v3f player_pos, v3f player_dir, v3f cam_pos, v3f cam_d
      ^
minetest/src/client/mumble.cpp: In function ‘void step(v3f, v3f, v3f, v3f)’:
minetest/src/client/mumble.cpp:118:6: error: ‘lm’ was not declared in this scope
  if (lm == NULL)
      ^
minetest/src/client/mumble.cpp:121:6: error: ‘lm’ was not declared in this scope
  if (lm->version != 2) {
      ^
minetest/src/client/mumble.cpp:126:2: error: ‘lm’ was not declared in this scope
  lm->tick++;
  ^
```

**Comment by NathanSalapat** [comment]
I'd love to test this, but how would I build it. Can somebody point me in the right direction as how to build minetest with a PR that isn't part of core. I know how to build already, just not how to add this patch to the build process.

**Comment by PilzAdam** [comment]
@NathanSalapat download @ShadowNinja's complete branch from here: https://github.com/ShadowNinja/minetest/archive/mumble.zip

**Comment by NathanSalapat** [comment]
Will try it out, thanks.

**Comment by sofar** [comment]
compiles, but segfaults as soon as game steps are run with:


```
Thread 1 (Thread 0x7ffff7f9b800 (LWP 24399)):
#0  0x00007ffff31f6bd0 in wcsncpy () from /usr/lib/libc.so.6
No symbol table info available.
#1  0x00000000004bdbc9 in MumbleManager::step (this=this@entry=0x7fffffffc778, player_pos=..., player_dir=..., cam_pos=..., cam_dir=...) at /home/sofar/git/minetest/src/client/mumble.cpp:123
No locals.
#2  0x000000000066a19d in Game::run (this=this@entry=0x7fffffffc6f0) at /home/sofar/git/minetest/src/game.cpp:1909
        cam_smoothing = <optimized out>
        player = 0x19939f0
        graph = {m_log = std::deque with 1 elements = {{values = std::map with 5 elements = {["client_received_packets"] = 137, ["mainloop_draw"] = 0.00200000009, ["mainloop_dtime"] = 17.8950005, ["mainloop_other"] = 17.8950005, 
                ["packets_lost"] = 0}}}, m_log_max_size = 200}
        stats = {drawtime = 2, beginscenetime = 1, endscenetime = 0, dtime_jitter = {max = 0, min = 0, avg = 0.715800047, counter = 14.8950005, max_sample = 17.1792011, min_sample = 0, max_fraction = 23.9665184}, busy_time_jitter = {max = 0, min = 0, 
            avg = 357.899994, counter = 14.8950005, max_sample = 17537.0996, min_sample = 0, max_fraction = 0}}
        cam_view_target = {camera_yaw = 0, camera_pitch = 0}
        cam_view = {camera_yaw = 0, camera_pitch = 0}
        runData = {dig_index = 0, new_playeritem = 0, pointed_old = {type = POINTEDTHING_NOTHING, node_undersurface = {X = 0, Y = 0, Z = 0}, node_abovesurface = {X = 0, Y = 0, Z = 0}, object_id = -1}, digging = false, ldown_for_dig = false, 
          left_punch = false, update_wielded_item_trigger = false, reset_jump_timer = false, nodig_delay_timer = -17.8950005, dig_time = 0, dig_time_complete = 0, repeat_rightclick_timer = 0, object_hit_delay_timer = -17.8950005, 
          time_from_last_punch = 27.8950005, selected_object = 0x0, jump_timer = 0, damage_flash = 0, update_draw_list_timer = 0, statustext_time = 0, fog_range = 180, update_draw_list_last_cam_dir = {X = 0, Y = 0, Z = 1}, profiler_current_page = 0, 
          profiler_max_page = 3, time_of_day = 0.967666686, time_of_day_smooth = 0.0483833365}
        draw_times = {last_time = 32483, busy_time = 17895, sleep_time = 0}
        flags = {invert_mouse = false, show_chat = true, show_hud = true, show_minimap = true, force_fog_off = false, show_debug = false, show_profiler_graph = false, disable_camera_update = false, first_loop_after_window_activation = false, 
          camera_offset_changed = false}
        dtime = 17.8950005
```

**Comment by ENDev15** [comment]
I'll try building it later if I get a change, and just see what happens @ShadowNinja.

**Comment by ShadowNinja** [comment]
@Elinvention: Looks like CMake isn't invalidating its cache correctly and is adding `src/client/mumble.cpp` even when USE_MUMBLE is false -- or USE_MUMBLE isn't being added to `src/cmake_config.h` like it should.

**Comment by ShadowNinja** [comment]
@sofar: Try now, I think I've fixed that.

**Comment by sofar** [comment]
will try to test tonight.

**Comment by Elinvention** [comment]
After I had to manually add `USE_MUMBLE`, `PROJECT_NAME_W` and `PROJECT_NAME_CW` to `src/cmake_config.h` because cmake doesn't want to do his job, it finally compiled.

To test it, I fed mumble on my notebook with music through pulseaudio while moving the player on my pc. It works, but there are some issues. Depending on the angle between the two players Mumble plays not positionally. Also the game crashed sometimes immediately after loading screen and to make it work again I had to restart Mumble. I collected a core dump and this backtrace:

```
#0  __wcsncpy (
    dest=0x7f27241e6000 <error: Cannot access memory at address 0x7f27241e6000>, src=<optimized out>, n=275) at wcsncpy.c:81
        c = 0
        s = 0x7f27241e5450 <error: Cannot access memory at address 0x7f27241e5450>
#1  0x00000000004bfa3b in MumbleManager::start(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) ()
No symbol table info available.
#2  0x00000000006f295c in Game::startup(bool*, bool, InputHandler*, irr::IrrlichtDevice*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, unsigned short, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, bool*, ChatBackend*, SubgameSpec const&, bool) ()
No symbol table info available.
#3  0x00000000006f91e2 in the_game(bool*, bool, InputHandler*, irr::IrrlichtDevice*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::c---Type <return> to continue, or q <return> to quit---
d::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned short, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, ChatBackend&, bool*, SubgameSpec const&, bool) ()
No symbol table info available.
#4  0x00000000004bc9a2 in ClientLauncher::run(GameParams&, Settings const&) ()
No symbol table info available.
#5  0x00000000004b46f8 in main ()
No symbol table info available.
```

**Comment by ShadowNinja** [comment]
@Elinvention: I fixed a few bugs in `MumbleManager::start` including a wcsncpy one, try now.

**Comment by Elinvention** [comment]
This are the issues I found:

- if you start a server from the GUI, the context of the player from the server is set to `|30000`, while others have `<ip>|30000`.
- if you connect with a client on 127.0.0.1, the context is `127.0.0.1|30000|30000`?!?
- vectors are 10x, so you have to be really close to hear something from the other end. I think you should divide them by `BS`, or even more (add a setting for this?).

I think IP is not a good context, for example if the server is behind a NAT, LAN players and WAN players will have different context. I propose to use the world name as context.

To get those values I wrote a small C program that reads from shared memory (github doesn't let me attach it because file type is not supported... why?).

**Comment by ShadowNinja** [comment]
Only the server knows what the worldname is, so that won't work.  I can make the address fall back to server_address I suppose, but I don't really know what a better context would be.  We could just use a `mumble_context` setting of course.  The position multiplication can be easily fixed and I'll push it in a moment.  The `127.0.0.1|30000|30000` thing sounds really bizarre, and I hove no idea how that'd happen.

**Comment by sofar** [comment]
Can't you use something like e.g. a hash of the server seed and something each connected client knows about the server, like the motd?

**Comment by est31** [comment]
This brings us back to the good old "how to uniquely identify servers" debate.

**Comment by ShadowNinja** [comment]
Using the seed is a good idea.  It's not necessarily unique though.  I've seen servers where the seed is set to 0 or a similarly non-random value.

**Comment by ENDev15** [comment]
Built fine using:
```
cmake . -DENABLE_GETTEXT=1 -DENABLE_FREETYPE=1 -DENABLE_LEVELDB=1; make -j$(nproc); sudo make install; minetest;
```
However, my computer repeatedly froze 30% of the way through the CMake build process until I ran `cmake .` by itself then the above.

During the build, `mumble.cpp.o` was mentioned, so I assume it built properly. Minetest also launches no differently and with no errors.

I can't figure out how to properly test it though. With Mumble open, I see no difference in Minetest even when connected to my Mumble server. Would me being the only person on the mumble server be the issue? Or did I not build it properly?

PS: Thanks for your work @ShadowNinja, really excited for this as I'm starting a let's play soon with a friend.

**Comment by ENDev15** [comment]
Doesn't seem to be functional (for me anyway). No overlay, though I didn't try to configure it. Audio is played at all times no matter the position. I didn't use any extra options during build or after the fact, just built (with the commands below), launched, and tested. 
```
cmake .
cmake . -DENABLE_GETTEXT=1 -DENABLE_FREETYPE=1 -DENABLE_LEVELDB=1
make -j$(nproc)
sudo make install
```
Game launches normally, with no weird errors in the log (other than the fact that technic can't find any of its mesh files >_<).

Not sure if I miss configured something, or what, but it doesn't seem to be working for me.

**Comment by Elinvention** [comment]
@ENDev15 as I said previously there is a problem with contexts. You have to start a server and connect to that server using the same IP address for all the clients. Otherwise they will have different contexts and positional audio won't work.

**Comment by ShadowNinja** [comment]
@ENDev15: On Linux you have to run it with the mumble preload library preloaded.  The easiest way to do this is through the mumble-overlay binary.  Run `mumble-overlay minetest` while mumble is running to get the overlay to work.

**Comment by davidebeatrici** [comment]
Hi,

Thank you for your interest in Mumble's positional audio system!

If you want to test the functionality of your integration, you can use [MumblePAHelper](https://github.com/mumble-voip/mumble-pahelper).
You need Qt to build it, but I can provide you an archive with the binary, if you want.

For any questions feel free to ask me.

**Comment by paramat** [comment]
Rubenwardy feels this is possibly not worth dev time http://irc.minetest.net/minetest-dev/2017-08-12#i_5041779
I am :-1: as very few people use chat audio in Minetest and i feel the feature will be rarely used and is too specialist to be worth the dev time to maintain. Doesn't seem to work currently.
I think your dev time is better spent on higher priorities, like your many other unfinished PRs.

**Comment by nerzhul** [comment]
i agree with closing this, it's a new and very rare feature to maintain
